src/QsCompiler/QirGeneration/AddOutputRecording.cs

```diff
@@ -0,0 +1,572 @@
+            }
+
+            public override QsCompilation OnCompilation(QsCompilation compilation)
+            {
+                compilation = base.OnCompilation(compilation);
+                return new QsCompilation(compilation.Namespaces, this.SharedState.EntryPointNames.ToImmutable());
+            }
+
+            private class NamespaceTransformation : NamespaceTransformation<TransformationState>
+            {
+                private OutputRecorder record;
+                private readonly bool alwaysCreateWrapper;
+
+                public NamespaceTransformation(SyntaxTreeTransformation<TransformationState> parent, OutputRecorder recorder, bool alwaysCreateWrapper)
+                    : base(parent)
+                {
+                    this.record = recorder;
+                    this.alwaysCreateWrapper = alwaysCreateWrapper;
+                }
+
+                private QsCallable CreateEntryPointWrapper(QsCallable c)
+                {
+                    QsScope CreateWrapperBody()
+                    {
+                        var (argType, returnType) = (c.Signature.ArgumentType, c.Signature.ReturnType);
+                        var callableType = c.Kind.IsOperation
+                            ? ResolvedTypeKind.NewOperation(Tuple.Create(argType, returnType), c.Signature.Information)
+                            : ResolvedTypeKind.NewFunction(argType, returnType);
+                        var callee = SyntaxGenerator.GlobalCallable(c.FullName, callableType, QsNullable<ImmutableArray<ResolvedType>>.Null);
+                        var callArgs = SyntaxGenerator.ArgumentTupleAsExpression(c.ArgumentTuple);
+
+                        var (outputTuple, variableDeclarations) = CreateDeconstruction(c.Signature.ReturnType);
+                        var outputDeconstruction = new QsStatement(
+                            QsStatementKind.NewQsVariableDeclaration(new QsBinding<TypedExpression>(
+                                QsBindingKind.ImmutableBinding,
+                                outputTuple,
+                                SyntaxGenerator.CallNonGeneric(callee, callArgs))),
+                            new LocalDeclarations(variableDeclarations.Values.ToImmutableArray()),
+                            QsNullable<QsLocation>.Null,
+                            QsComments.Empty);
+
+                        var paramDecl = SyntaxGenerator.ExtractItems(c.ArgumentTuple).ValidDeclarations();
+                        var localDeclarations = variableDeclarations.Values.Concat(paramDecl)
+                            .ToImmutableDictionary(decl => decl.VariableName);
+                        return new QsScope(
+                            this.record(outputTuple, localDeclarations).Prepend(outputDeconstruction).ToImmutableArray(),
+                            new LocalDeclarations(paramDecl));
+                    }
+
+                    var wrapperName = new QsQualifiedName(c.FullName.Namespace, c.FullName.Name + "__main");
```

Comment:
Please use the `NameDecorator` class to generate names for callables and variables. If the behavior you are looking for is not in the `NameDecorator`, then please add it to the `NameDecorator`. That way, when we go to update our naming standards, this generated name doesn't accidentally get left behind, and if you add new behavior to the `NameDecorator`, this can be considered and incorporated into new standards we come up with for generated names.

Reply:
As discussed offline, I can't. This name has to be predictable due to being an entry point. I don't like it as is, and I'll think if I can come up with something cleaner (hence the todo in the code).

Reply:
Mimicking the SpecializationSuffix in InferTargetInstructions, I decided to make this an argument to the AddOutputRecording that is defined by the QIR generation, cleaning up (meaning removing) the hardcoded dependency.